src/lib/Parser.hs

```diff
@@ -267,34 +268,49 @@ addImplicitImplicitArgs sourcePos (Just typ) ex =
 interfaceDef :: Parser [UDecl]
 interfaceDef = do
   keyWord InterfaceKW
-  -- TODO: allow multi-parameter type classes
-  interfaceName <- upperName <|> symName
-  (oneAnnBinder, pos) <- withPos annBinder
+  (tyCon, pos) <- withPos tyConDef
   keyWord WhereKW
-  record <- withSrc $ URecordTy <$> interfaceRecordFields ":"
-  let consName = mkInterfaceConsName interfaceName
-  let (varTypeName, oneAnnBinder') = case oneAnnBinder of
-        Bind v    -> (varName v, oneAnnBinder)
-        Ignore ty -> let genVar = mkName $ mkNoShadowingStr "gen" in
-            (genVar, Bind (genVar:>ty))
-  let tyCon = UConDef interfaceName $ toNest [oneAnnBinder']
-  let funDefs = mkFunDefs (pos, varTypeName, interfaceName) record
-  return $ (UData tyCon [UConDef consName (toNest [Ignore record])]) : funDefs
+  recordFields <- withSrc $ interfaceRecordFields ":"
+  let (UConDef interfaceName uAnnBinderNest) = tyCon
+      record = (URecordTy . NoExt) <$> recordFields
+      consName = mkInterfaceConsName interfaceName
+      ((_, varNames), uAnnBinderNest') =
+          mapAccumR mkFreeVarTypeName (0, []) uAnnBinderNest
+      tyCon' = UConDef interfaceName uAnnBinderNest'
+      funDefs = mkFunDefs (pos, varNames, interfaceName) recordFields
+  return $ (UData tyCon' [UConDef consName (toNest [Ignore record])]) : funDefs
   where
-    mkFunDefs :: (SrcPos, Name, Name) -> UExpr -> [UDecl]
-    mkFunDefs meta ~(WithSrc _ (URecordTy (NoExt (LabeledItems items)))) =
+    mkFreeVarTypeName
+      :: (Int, [Name]) -> UAnnBinder -> ((Int, [Name]), UAnnBinder)
+    mkFreeVarTypeName (n, varNames) (Ignore ty) =
+      let genVarName = mkName $ mkNoShadowingStr ("gen" ++ show n) in
+      ((n + 1, genVarName:varNames), Bind (genVarName:>ty))
+    mkFreeVarTypeName (n, varNames) b@(Bind v) = ((n, (varName v):varNames), b)
+    -- From an interface
+    --   interface I a:Type b:Type where
+    --     f : a -> b
+    -- mkFunDefs generates the equivalent of the following function definition:
+    --   def f (instance# : I a b) ?=> a -> b =
+    --     (I# {f=f,...}) = instance#
+    --     f
+    -- where I# is an automatically generated constructor of I.
+    mkFunDefs
+      :: (SrcPos, [Name], Name) -> WithSrc (LabeledItems UExpr) -> [UDecl]
+    mkFunDefs meta ~(WithSrc _ (LabeledItems items)) =
```

Comment:
Why use `WithSrc` if you immediately throw away the source position? Could just be
```
:: (SrcPos, [Name], Name) -> LabeledItems UExpr -> [UDecl]
```
right?

Reply:
Yes, I guess I was just avoiding creating a new variable to hold the `recordFields` from the `WithSrc`, but having a cleaner type is worth the trouble :)